nx/c_src/pytorch.cpp

```diff
@@ -0,0 +1,327 @@
+#include <ATen/ATen.h>
+#include <iostream>
+
+#include "nx_nif_utils.hpp"
+
+ErlNifResourceType *TENSOR_TYPE;
+
+std::map<const std::string, const at::ScalarType> dtypes = {{"byte", at::kByte}, {"char", at::kChar}, {"short", at::kShort}, {"int", at::kInt}, {"long", at::kLong}, {"half", at::kHalf}, {"float", at::kFloat}, {"double", at::kDouble}};
+std::map<const std::string, const int> dtype_sizes = {{"byte", 1}, {"char", 1}, {"short", 2}, {"int", 4}, {"long", 8}, {"half", 2}, {"float", 4}, {"double", 8}};
+
+ERL_NIF_TERM create_tensor_resource(ErlNifEnv *env, at::Tensor tensor)
+{
+  ERL_NIF_TERM ret;
+  at::Tensor *tensorPtr;
+
+  tensorPtr = (at::Tensor *)enif_alloc_resource(TENSOR_TYPE, sizeof(at::Tensor));
+  if (tensorPtr == NULL)
+    return enif_make_badarg(env);
+
+  new (tensorPtr) at::Tensor(tensor.variable_data());
+
+  ret = enif_make_resource(env, tensorPtr);
+  enif_keep_resource(tensorPtr);
```

Comment:
You don't need this keep reosurce call. All `enif_make_resource` are by default automatically kept, so if you have this additional `enif_keep_resource` call, it means it won't ever be garbage collected.